File: ts/src/core/dev_schema.ts
import { createHash } from "crypto";
import * as fs from "fs";
import * as path from "path";
import type { RuntimeDevSchemaConfig } from "./config";

const STATE_DIR = ".ent";
const STATE_FILE = "dev_schema.json";
const DEFAULT_SCHEMA_DIR = path.join("src", "schema");
const DEFAULT_SCHEMA_PREFIX = "ent_dev";
const MAX_SCHEMA_LEN = 63;

export interface ResolvedDevSchema {
  enabled: boolean;
  schemaName?: string;
  branchName?: string;
  includePublic?: boolean;
}

interface DevSchemaState {
  schemaName: string;
  branchName?: string;
  includePublic?: boolean;
  ignoreBranches?: string[];
}

export function resolveDevSchema(
  cfg?: RuntimeDevSchemaConfig,
): ResolvedDevSchema {
  if (!isDevSchemaEnabled(cfg)) {
    return { enabled: false };
  }

  if (cfg?.schemaName) {
    const schemaName = sanitizeIdentifier(cfg.schemaName);
    return {
      enabled: true,
      schemaName,
      includePublic: cfg.includePublic === true,
    };
  }

  if (cfg) {
    const branchName = requireCurrentBranch(
      "dev branch schemas are enabled but the current git branch could not be determined. Set devSchema.schemaName explicitly.",
    );
    return {
      enabled: true,
      schemaName: buildSchemaName(branchName),
      branchName,
      includePublic: cfg.includePublic === true,
    };
  }

  const state = loadDevSchemaState(resolveStatePath());
  if (state?.schemaName) {
    const branchName = state.branchName;
    if (branchName) {
      // State-file mode is tied to the branch that produced the generated
      // schema metadata. Fail closed after branch switches until codegen
      // refreshes the state file.
      const currentBranch = requireCurrentBranch(
        `dev branch schema state was generated for branch "${branchName}" but the current git branch could not be determined. Run ent codegen to regenerate or set devSchema.schemaName explicitly.`,
      );
      if (currentBranch !== branchName) {
        throw new Error(
          `dev branch schema state was generated for branch "${branchName}" but current branch is "${currentBranch}". Run ent codegen to regenerate or set devSchema.schemaName explicitly.`,
        );
      }
    }

    return {
      enabled: true,
      schemaName: sanitizeIdentifier(state.schemaName),
      branchName,
      includePublic: state.includePublic === true,
    };
  }

  const branchName = requireCurrentBranch(
    "dev branch schemas are enabled but the current git branch could not be determined. Set devSchema.schemaName explicitly or run ent codegen to regenerate src/schema/.ent/dev_schema.json.",
  );
  return {
    enabled: true,
    schemaName: buildSchemaName(branchName),
    branchName,
    includePublic: false,
  };
}

export function isDevSchemaEnabled(cfg?: RuntimeDevSchemaConfig): boolean {
  const nodeEnv = (process.env.NODE_ENV || "").toLowerCase();
  if (nodeEnv === "production") {
    return false;
  }

  const envEnabled = parseEnvBool("ENT_DEV_SCHEMA_ENABLED");
  if (envEnabled !== undefined) {
    return envEnabled;
  }
  if (cfg) {
    if (cfg.enabled !== true) {
      return false;
    }
    const branch = resolveGitBranch();
    if (isBranchIgnored(cfg.ignoreBranches, branch)) {
      return false;
    }
    return true;
  }
  const state = loadDevSchemaState(resolveStatePath());
  if (!state?.schemaName) {
    return false;
  }
  const branch = resolveGitBranch();
  if (isBranchIgnored(state.ignoreBranches, branch)) {
    return false;
  }
  return true;
}

function loadDevSchemaState(statePath?: string): DevSchemaState | undefined {
  if (!statePath || !fs.existsSync(statePath)) {
    return undefined;
  }
  const raw = fs.readFileSync(statePath, "utf8");
  let data: DevSchemaState;
  try {
    data = JSON.parse(raw) as DevSchemaState;
  } catch (err) {
    throw new Error(`invalid dev schema state file at ${statePath}`);
  }
  if (!data || !data.schemaName) {
    return undefined;
  }
  return data;
}

function requireCurrentBranch(message: string): string {
  const branch = resolveGitBranch();
  if (!branch) {
    throw new Error(message);
  }
  return branch;
}

function slugify(input: string): string {
  if (!input) {
    return "";
  }
  const lower = input.toLowerCase();
  let out = "";
  let lastUnderscore = false;
  for (const ch of lower) {
    const isAlpha = ch >= "a" && ch <= "z";
    const isDigit = ch >= "0" && ch <= "9";
    if (isAlpha || isDigit) {
      out += ch;
      lastUnderscore = false;
      continue;
    }
    if (!lastUnderscore) {
      out += "_";
      lastUnderscore = true;
    }
  }
  return out.replace(/^_+|_+$/g, "");
}

function sanitizeIdentifier(input: string): string {
  const slug = slugify(input);
  if (!slug) {
    return "schema";
  }
  let normalized = slug;
  if (normalized[0] >= "0" && normalized[0] <= "9") {
    normalized = `schema_${normalized}`;
  }
  if (normalized.length > MAX_SCHEMA_LEN) {
    return normalized.slice(0, MAX_SCHEMA_LEN);
  }
  return normalized;
}

function shortHash(input: string): string {
  return createHash("sha1").update(input).digest("hex").slice(0, 8);
}

function buildSchemaName(branch: string): string {
  const prefix = sanitizeIdentifier(DEFAULT_SCHEMA_PREFIX);
  let branchSlug = slugify(branch);
  if (!branchSlug) {
    branchSlug = "branch";
  }
  const hash = shortHash(branch);
  let name = [prefix, branchSlug, hash].join("_");
  if (name.length <= MAX_SCHEMA_LEN) {
    return name;
  }

  const over = name.length - MAX_SCHEMA_LEN;
  if (over > 0 && branchSlug.length > 1) {
    branchSlug = branchSlug.slice(
      0,
      branchSlug.length - Math.min(over, branchSlug.length - 1),
    );
  }
  name = [prefix, branchSlug, hash].join("_");
  if (name.length > MAX_SCHEMA_LEN) {
    return name.slice(0, MAX_SCHEMA_LEN);
  }
  return name;
}

function resolveStatePath(): string | undefined {
  const start = process.cwd();
  const root = findGitRoot(start) || start;
  const schemaDir = path.join(root, DEFAULT_SCHEMA_DIR);
  return path.join(schemaDir, STATE_DIR, STATE_FILE);
}

function isBranchIgnored(
  ignoreBranches: string[] | undefined,
  branch: string,
): boolean {
  if (!branch || !ignoreBranches || ignoreBranches.length === 0) {
    return false;
  }
  for (const name of ignoreBranches) {
    if (!name || !name.trim()) {
      continue;
    }
    if (name === branch) {
      return true;
    }
  }
  return false;
}

function parseEnvBool(key: string): boolean | undefined {
  const raw = process.env[key];
  if (!raw) {
    return undefined;
  }
  const val = raw.trim().toLowerCase();
  if (["1", "true", "t", "yes", "y"].includes(val)) {
    return true;
  }
  if (["0", "false", "f", "no", "n"].includes(val)) {
    return false;
  }
  return undefined;
}

function resolveGitBranch(): string {
  const start = process.cwd();
  const root = findGitRoot(start);
  if (!root) {
    return "";
  }
  const gitDir = resolveGitDir(path.join(root, ".git"));
  if (!gitDir) {
    return "";
  }
  const headPath = path.join(gitDir, "HEAD");
  if (!fs.existsSync(headPath)) {
    return "";
  }
  const head = fs.readFileSync(headPath, "utf8").trim();
  if (!head.startsWith("ref:")) {
    return "";
  }
  return head
    .replace("ref:", "")
    .trim()
    .replace(/^refs\/heads\//, "");
}

function resolveGitDir(gitPath: string): string | undefined {
  if (!fs.existsSync(gitPath)) {
    return undefined;
  }
  const stat = fs.statSync(gitPath);
  if (stat.isDirectory()) {
    return gitPath;
  }
  const contents = fs.readFileSync(gitPath, "utf8").trim();
  if (!contents.startsWith("gitdir:")) {
    return undefined;
  }
  let dir = contents.replace("gitdir:", "").trim();
  if (!path.isAbsolute(dir)) {
    dir = path.join(path.dirname(gitPath), dir);
  }
  return dir;
}

function findGitRoot(start: string): string | undefined {
  let dir = start;
  while (true) {
    const gitPath = path.join(dir, ".git");
    if (fs.existsSync(gitPath)) {
      return dir;
    }
    const parent = path.dirname(dir);
    if (parent === dir) {
      return undefined;
    }
    dir = parent;
  }
}

File: internal/devschema/devschema.go
package devschema

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

const (
	DefaultPrefix    = "ent_dev"
	DefaultPruneDays = 30
	MaxSchemaLen     = 63
)

type Config struct {
	Enabled        bool
	SchemaName     string
	IncludePublic  bool
	IgnoreBranches []string
	PruneEnabled   bool
	PruneDays      int
}

type Options struct {
	NodeEnv    string
	RepoRoot   string
	SchemaPath string
}

type Result struct {
	Enabled        bool
	SchemaName     string
	BranchName     string
	IncludePublic  bool
	IgnoreBranches []string
	PruneEnabled   bool
	PruneDays      int
}

func Resolve(cfg *Config, opts Options) (*Result, error) {
	nodeEnv := opts.NodeEnv
	if nodeEnv == "" {
		nodeEnv = os.Getenv("NODE_ENV")
	}
	if strings.EqualFold(nodeEnv, "production") {
		return &Result{Enabled: false}, nil
	}

	var state *State
	if cfg == nil {
		var err error
		state, err = LoadState(opts)
		if err != nil {
			return nil, err
		}
	}

	envEnabled := parseEnvBool("ENT_DEV_SCHEMA_ENABLED")
	enabled := false
	if envEnabled != nil {
		enabled = *envEnabled
	} else if cfg != nil {
		enabled = cfg.Enabled
	} else if state != nil && state.SchemaName != "" {
		enabled = true
	}

	if !enabled {
		return &Result{Enabled: false}, nil
	}

	currentBranch := ""
	currentBranchResolved := false
	resolveCurrentBranch := func() string {
		if !currentBranchResolved {
			currentBranch = ResolveGitBranch(opts.RepoRoot)
			currentBranchResolved = true
		}
		return currentBranch
	}

	includePublic := false
	if cfg != nil {
		includePublic = cfg.IncludePublic
	} else if state != nil {
		includePublic = state.IncludePublic
	}
	ignoreBranches := []string{}
	if cfg != nil && len(cfg.IgnoreBranches) > 0 {
		ignoreBranches = cfg.IgnoreBranches
	} else if state != nil && len(state.IgnoreBranches) > 0 {
		ignoreBranches = state.IgnoreBranches
	}

	schemaName := ""
	explicitSchema := false
	branchName := ""
	pruneEnabled, pruneDays := resolvePrune(cfg, state)

	if cfg != nil && cfg.SchemaName != "" {
		schemaName = cfg.SchemaName
		explicitSchema = true
	} else if state != nil {
		schemaName = state.SchemaName
		branchName = state.BranchName
		if schemaName != "" && branchName == "" {
			explicitSchema = true
		}
	}

	if explicitSchema {
		schemaName = sanitizeIdentifier(schemaName)
	} else if cfg != nil {
		branchName = resolveCurrentBranch()
		if branchName == "" {
			return nil, fmt.Errorf("dev branch schemas are enabled but the current git branch could not be determined; set devSchema.schemaName explicitly")
		}
		schemaName = buildSchemaName(branchName)
	} else {
		if schemaName != "" && branchName != "" {
			// State-file mode is tied to the branch that generated the schema.
			// Fail closed after branch switches until codegen refreshes the state.
			current := resolveCurrentBranch()
			if current == "" {
				return nil, fmt.Errorf("dev branch schema state was generated for branch %q but the current git branch could not be determined; run ent codegen to regenerate or set devSchema.schemaName explicitly", branchName)
			}
			if current != branchName {
				return nil, fmt.Errorf("dev branch schema state was generated for branch %q but current branch is %q; run ent codegen to regenerate or set devSchema.schemaName explicitly", branchName, current)
			}
		}
		if schemaName == "" {
			branchName = resolveCurrentBranch()
			if branchName == "" {
				return nil, fmt.Errorf("dev branch schemas are enabled but the current git branch could not be determined; set devSchema.schemaName explicitly or run ent codegen to regenerate src/schema/.ent/dev_schema.json")
			}
			schemaName = buildSchemaName(branchName)
		}
	}
	branchForIgnore := branchName
	if branchForIgnore == "" {
		branchForIgnore = resolveCurrentBranch()
	}
	if envEnabled == nil && isBranchIgnored(ignoreBranches, branchForIgnore) {
		return &Result{Enabled: false}, nil
	}

	return &Result{
		Enabled:        true,
		SchemaName:     schemaName,
		BranchName:     branchName,
		IncludePublic:  includePublic,
		IgnoreBranches: ignoreBranches,
		PruneEnabled:   pruneEnabled,
		PruneDays:      pruneDays,
	}, nil
}

func WriteStateFromConfig(cfg *Config, opts Options) (*Result, error) {
	res, err := Resolve(cfg, opts)
	if err != nil {
		return nil, err
	}
	if err := WriteState(res, opts); err != nil {
		return nil, err
	}
	return res, nil
}

func resolvePrune(cfg *Config, state *State) (bool, int) {
	if cfg != nil {
		days := cfg.PruneDays
		if days <= 0 {
			days = DefaultPruneDays
		}
		return cfg.PruneEnabled, days
	}
	if state != nil {
		days := state.PruneDays
		if days <= 0 {
			days = DefaultPruneDays
		}
		return state.PruneEnabled, days
	}
	return false, DefaultPruneDays
}

func ResolveGitBranch(repoRoot string) string {
	start := repoRoot
	if start == "" {
		if cwd, err := os.Getwd(); err == nil {
			start = cwd
		}
	}
	root := findGitRoot(start)
	if root == "" {
		return ""
	}
	return branchFromGitDir(resolveGitDir(filepath.Join(root, ".git")))
}
